Updates to support source code linting

ehoogerbeets opened the pull request:
- added Parser.getType() method
- clarified some jsdocs
    - Rule.match() should return a Result instance, an array of
      Result instances, or undefined if no match
- added IntermediateRepresentation class to represent the results of
  parsing a file
- Parser.parse() should now return the intermediate representation
- Added support for logging provided by the lint tool so that plugins
  can use the linter's logging
    - getLogger() function passed to the constructors

ehoogerbeets committed:
Added method and updated docs

- added Parser.getType() method
- clarified some jsdocs
    - Rule.match() should return a Result instance, an array of
      Result instances, or undefined if no match
- Parser.parse() should now return the intermediate representation

ehoogerbeets committed:
Updates for the support of linting source code

ehoogerbeets committed:
Added release note items

ehoogerbeets committed:
Clarify docs for Rule.getRuleType()

ehoogerbeets committed:
Fix unit tests

ehoogerbeets commented:
Ready for review now. This PR has to go first, and then we can do the linter itself and a few other projects that depend on these.

ehoogerbeets committed:
Update version to 1.5.0

ehoogerbeets committed:
Updated the comments a bit

wadimw commented on src/Plugin.js:
```diff
@@ -30,15 +32,26 @@ class Plugin {
+    /**
+     * Return the version of the API that this plugin was built for. If
+     */
+    getAPIVersion() {
```
IMO this duplicates functionality of dependency versioning in Node's `package.json`.

A custom plugin should have a dependency on a specific version of i18nlint-common, which itself is a dependency of ilib-lint. Whenever common's API changes in a breaking way, i18nlint-common's version should reflect that by bumping its major number. So, the custom plugin dependent on the old common will not be allowed to use the new common until it's fixed accordingly - same goes for ilib-lint.

wadimw commented on src/Plugin.js:
```diff
@@ -30,15 +32,26 @@ class Plugin {
     init() {}
 
+    /**
+     * Return the version of the API that this plugin was built for. If
```
Cut comment

wadimw commented on src/Plugin.js:
```diff
@@ -30,15 +32,26 @@ class Plugin {
      * Initialize the current plugin, if necessary.
      *
      * @abstract
+     * @returns {Promise|undefined} a promise to initialize or undefined if the
```
It should specifically be a `Promise<void>` to indicate that no return value is expected.

ehoogerbeets replied:
done

wadimw commented on src/NullLogger.js:
```diff
@@ -0,0 +1,32 @@
+/**
+ * @class null logger with a similar API as log4js which produces no output
+ */
+class NullLogger {
```
I don't think NullLogger specifically should be part of this package.

Since logger is being provided by means of constructor dependency injection, the `Logger` interface becomes part of the overall API. Still, providing an actual *instance* of `Logger` is a responsibility of the app that wants to instantiate the dependency client (e.g. `Parser`).

To match the convention we already have (and because JS does not have interfaces), `NullLogger` could instead become an abstract class
```js
/** @abstract */
class Logger {
    /** @abstract */
    fatal(/** @type {string} */ message) {}
    
    // etc
}
```

`NullLogger` could then be a class only used for testing purposes
```js
class NullLogger extends Logger {
    /** @override */
    fatal(message) { /* no-op */ }
    // etc
}
```

while `ilib-lint` could plug in an own `Logger` instance which adapts log4js.

ehoogerbeets replied:
removed

wadimw commented on src/Parser.js:
```diff
@@ -32,6 +34,9 @@ class Parser {
         }
+        if (!options) return;
+        this.API = options.API;
+        this.logger = (this.API && this.API.getLogger()) || new NullLogger();
```
Logger should not be instantiated in here. If this dependency is optional, instead of constructor injection we could use a setter injection `setLogger(/** @type {Logger} */ logger)` and define `logger` property as `Logger | undefined` so that implementation of `Parser` would know that it should call `logger?.error("some message")`.

ehoogerbeets replied:
removed

wadimw commented on src/IntermediateRepresentation.js:
```diff
@@ -0,0 +1,80 @@
+            }
+            return true;
+        });
+        // logger.trace("params is " + JSON.stringify(params));
```
leftover comment

ehoogerbeets replied:
removed

wadimw commented on src/Rule.js:
```diff
@@ -75,29 +79,36 @@ class Rule {
-        // default rule type. If your rule is different, override this method.
-        return "line";
+        // default representation type. If your rule is different, override this method.
+        return "string";
```
There should not be a default expected representation type, because there is no default rule implementation.

wadimw commented on src/Parser.js:
```diff
@@ -72,14 +77,56 @@ class Parser {
+    getType() {}
+
+    /**
+     * For a "resource" type of parser, this returns a list of Resource instances
```
This method should be removed from the interface `Parser`, since most of them won't return any resources

ehoogerbeets replied:
removed

wadimw commented on src/IntermediateRepresentation.js:
```diff
@@ -0,0 +1,80 @@
+
+/**
+ * @class Representation of parser results
+ * @abstract
```
Perhaps instead of an abstract class, it could be concrete but generic `IntermediateRepresentation<T>`?

ehoogerbeets replied:
Actually, that was an artifact left over from when I copied the contents of a file from a different file. I removed the "@abstract" tag.

wadimw commented on src/Parser.js:
```diff
@@ -32,6 +34,9 @@ class Parser {
             throw new Error("Cannot instantiate abstract class Plugin directly!");
         }
+        if (!options) return;
+        this.API = options.API;
```
There is no definition of the `API` object interface in this package.

Additionally, I think it's unnecessary to store dependencies nested in `API` property.

ehoogerbeets replied:
The logger is actually different based on the parameter to it..  `getLogger("my.unique.logging.id")`
Once the plugins call this.API.getLogger, it should store the results of that call.

I added the LintAPI class to document what the API is... it only includes the call back for the getLogger so far. Maybe we should just provide the getLogger function in the options object for the constructor? That way, the constructor for each plugin can call getLogger with its unique id to get its own logger instance, and then we don't need the API that way...

ehoogerbeets replied:
Okay, just removed the LintAPI again and pass the logger in to the constructors instead

ehoogerbeets committed:
Addressed some of Wadim's review comments

- updated jsdoc documentation for many methods/classes
- removed NullLogger. Subclasses will have to get their own logger
  from the API if they need it.
- added LintAPI class to document what the API object is
- removed getResources() method

ehoogerbeets committed:
Remove API object

- pass the getLogger function to constructors directly
- updated documentation

ehoogerbeets committed:
Remove the getAPIVersion method

wadimw commented on README.md:
```diff
@@ -39,6 +39,19 @@ limitations under the License.
 
 ## Release Notes
 
+### v1.5.0
```
Forgot about one thing, shouldn't we bump major verison? I think the changes we make are breaking (like the type of object returned from Parser etc)

ehoogerbeets replied:
ok  updated

ehoogerbeets committed:
Update the major version
